### PR TITLE
Fixed #25143 -- Added ArrayField.from_db_value()

### PR DESCRIPTION
**Pull request by karanlyons**
See https://code.djangoproject.com/ticket/25143 and https://github.com/django/django/pull/5035#issuecomment-166167796.

### CODE REVIEW

File: tests/postgres_tests/models.py [hunk 1]
```diff
@@ -6,6 +6,32 @@
 )
 
 
+class Tag(object):
+    def __init__(self, tag_id):
+        self.tag_id = tag_id
+
+    def __eq__(self, other):
+        return type(other) is Tag and self.tag_id == other.tag_id
```
Comment (charettes) [inline tests/postgres_tests/models.py]: Nitpick but use `isinstance(other, Tag)`.

**Comment by charettes** [comment]
Could you rebase on top of the latest master to resolve the merge conflicts?

**Comment by karanlyons** [comment]
isort's complaining that the imports in `tests/postgres_tests/test_array.py` are incorrectly sorted, but I don’t believe they are, or at the very least the sorting issue should not have been introduced by this patch.

File: tests/postgres_tests/test_array.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 from .models import (
     ArrayFieldSubclass, CharArrayModel, DateTimeArrayModel, IntegerArrayModel,
     NestedIntegerArrayModel, NullableIntegerArrayModel, OtherTypesArrayModel,
-    PostgreSQLModel,
+    PostgreSQLModel, Tag
```
Comment (timgraham) [inline tests/postgres_tests/test_array.py]: need a trailing comma to appease isort style

File: docs/_theme/djangodocs-epub/static/docicons-warning.png [hunk 3]
```diff
@@ -1 +0,0 @@
-../../djangodocs/static/docicons-warning.png
```
Comment (timgraham) [inline docs/_theme/djangodocs-epub/static/docicons-warning.png]: revert these changes

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by karanlyons** [comment]
My apologies for the failed tests: My additions to the original pull were under the assumption that it was otherwise good to go.

There are two options for handling these failures, as they’re all related to the added `tags` attribute to `OtherTypesArrayModel`: We can either define `tags` in each test so that it’s not null, or allow null values in the model. I don’t rightly know what the preferred style is for Django’s tests, but it seems to be the latter. Is that correct?

**Comment by timgraham** [comment]
Looks like adding `Tag` values and testing the other locations the model is used wouldn't hurt.

**Comment by karanlyons** [comment]
This *should* be good to go now, but let me know if I’ve missed something.

File: django/contrib/postgres/fields/array.py [hunk 4]
```diff
@@ -102,6 +104,14 @@ def to_python(self, value):
             value = [self.base_field.to_python(val) for val in vals]
         return value
 
+    def _from_db_value(self, value, expression, connection, context):
+        if value is None:
```
Comment (timgraham) [inline django/contrib/postgres/fields/array.py]: I think we're missing a test for this condition (no tests fail if it's removed)

File: django/contrib/postgres/fields/array.py [hunk 5]
```diff
@@ -28,6 +28,8 @@ def __init__(self, base_field, size=None, **kwargs):
         if self.size:
             self.default_validators = self.default_validators[:]
             self.default_validators.append(ArrayMaxLengthValidator(self.size))
+        if hasattr(self.base_field, 'from_db_value'):
```
Comment (timgraham) [inline django/contrib/postgres/fields/array.py]: It would be helpful to add a comment like: "For performance, only add a from_db_value() method if the base field implements it."

**Comment by timgraham** [comment]
Addresses my two comments above and squash commits in https://github.com/django/django/pull/6279.
